### DISCUSSION TITLE
CVRP: CumulVar Constraint at last visit Node.

### DISCUSSION THREAD

**Discussion by burbot48** [google/or-tools#4046] [Routing (and legacy CP) questions]
Hello,

I try to develop a CVRP model. And I have to add a constraint for each Vehicle such the CumulVar of last node before returning the depot should be less than 4. Each Vehicle capacity is 8.
Could you please help to create this constraint.

Thank you.

**Comment by sschnug** [comment]
This probably depends a bit on what you are accumulating and which structure you can exploit.

(i'm assuming <= 4 here)

If it's time, it's probably non-decreasing:
- just change their upper-bound to 4 for all of them (but not for depot-nodes).

If it's capacity, it's probably not non-decreasing, but there might be a 0-transit on node -> depot:
- add a 4 to all node -> depot transits and together with your upper-bound of 8 this should be enough

If none of this make sense, it might be time for ugly and not as good propagating constraints making use of implications/reifications but i would avoid that.

**Reply by burbot48** [reply]
Thank you for your quick response.
What if I need to put a constraint for all nodes before the 2nd or 3rd last node?
How can I manage this situation in model?

**Reply by sschnug** [reply]
If your constraint depends on the absolute position in the tour, those efficient tricks probably won't work anymore.

In most cases you probably need to add a counting-dimension (one where each visit adds +1) and you then use general-purpose constraint-programming capabilities. But this propagation will cost a bit i would guess. 

I even ignored the fact that counting is accumulation from the front and 2nd/3rd last node will need to infer some accumulation from the end which might even further worsen propagation (as maybe something like: x = count-all; pos = x - counter(node) or something like that)

**Reply by burbot48** [reply]
Actually it depends on region definitions of nodes. If the last n visiting nodes in the vehicle are defined as same region. I have to put that constraint for all the nodes before last n. nodes in the vehicle . So It can be vary due to region definition.
